File: support/pagination/src/main/java/com/coffee/gu/Page.java
package com.coffee.gu;

import java.time.LocalDateTime;
import java.util.List;
import java.util.function.Function;

public record Page<T>(
        List<T> content,
        Boolean hasNext,
        LocalDateTime nextCursor,
        Long nextLastId
) {
    public static <E> Page<E> of(
            List<E> items,
            int pageSize,
            Function<E, LocalDateTime> cursorExtractor,
            Function<E, Long> idExtractor
    ) {
        boolean hasNext = items.size() > pageSize;

        List<E> content = hasNext
                ? items.subList(0, pageSize)
                : items;

        LocalDateTime nextCursor = null;
        Long nextLastId = null;

        if (hasNext && !content.isEmpty()) {
            E lastItem = content.getLast();
            nextCursor = cursorExtractor.apply(lastItem);
            nextLastId = idExtractor.apply(lastItem);
        }

        return new Page<>(content, hasNext, nextCursor, nextLastId);
    }

    public static <E> Page<E> of(List<E> items, int pageSize) {
        boolean hasNext = items.size() > pageSize;

        List<E> content = hasNext
                ? items.subList(0, pageSize)
                : items;

        return new Page<>(content, hasNext, null, null);
    }

    public <R> Page<R> map(Function<T, R> mapper) {
        List<R> mappedContent = this.content.stream().map(mapper).toList();
        return new Page<>(mappedContent, hasNext, nextCursor, nextLastId);
    }
}

File: support/pagination/src/main/kotlin/com/coffee/gu/Page.kt
package com.coffee.gu

import java.time.LocalDateTime

data class Page<T>(
    val content: List<T>,
    val hasNext: Boolean,
    val nextCursor: LocalDateTime? = null,
    val nextLastId: Long? = null,
) {

    fun <R> map(transform: (T) -> R): Page<R> {
        val mappedContent = this.content.map(transform)
        return Page(mappedContent, hasNext, nextCursor, nextLastId)
    }

    companion object {
        fun <E> of(
            items: List<E>,
            pageSize: Int,
            cursorExtractor: ((E) -> LocalDateTime)? = null,
            idExtractor: ((E) -> Long)? = null,
        ): Page<E> {
            val hasNext = items.size > pageSize
            val content = if (hasNext) items.subList(0, pageSize) else items

            var nextCursor: LocalDateTime? = null
            var nextLastId: Long? = null

            if (hasNext && content.isNotEmpty() && cursorExtractor != null && idExtractor != null) {
                val lastItem = content.last()
                nextCursor = cursorExtractor(lastItem)
                nextLastId = idExtractor(lastItem)
            }

            return Page(content, hasNext, nextCursor, nextLastId)
        }
    }
}